### PR TITLE
MinGW64+MSYS2 compilation issues

### PR DESCRIPTION
**Pull request by ostrovsky**

### CODE REVIEW

**Comment by ostrovsky** [comment]
I signed it!

File: src/google/protobuf/util/internal/utility.cc [hunk 1]
```diff
@@ -302,7 +302,7 @@ bool IsMap(const google::protobuf::Field& field,
 string DoubleAsString(double value) {
   if (value == std::numeric_limits<double>::infinity()) return "Infinity";
   if (value == -std::numeric_limits<double>::infinity()) return "-Infinity";
-  if (::isnan(value)) return "NaN";
+  if (std::isnan(value)) return "NaN";
```
Comment (xfxyjwf) [inline src/google/protobuf/util/internal/utility.cc]: std::isnan is a C++11 feature. We probably need to stick with ::isnan. Why can't "::isnan" be used here?

Comment (ostrovsky) [inline src/google/protobuf/util/internal/utility.cc]: Hello Feng Xiao,

I guess in MSYS2 with MinGW-w64 latest development environment (gcc 4.9.2)
C++11 is the default and this module fails to compile.  What is your
standard way of dealing with such portability issues?  Would some sort of
#ifdef __cplusplus be fine?

Best regards,

Karol

**Comment by nslay** [comment]
::isnan doesn't exist in VS 2008 or VS 2010. Well I read somewhere that you wished to support VS 2008 still. If you don't care about VS 2008/2010. Then disregard this message. The same goes for isfinite().

Could you use utility class in mathlimits.h to test for nan or finite? One could also just test the version of _MSC_VER and declare these otherwise standard functions as inline.

I use something like this in common.h in my own fork:
// VS 2010 and below do not have isnan() or isfinite()
#if defined(_MSC_VER) && _MSC_VER <= 1600
namespace google {
namespace protobuf {
// Place in namespace to protect against external projects with similar workarounds
// These are standard in C++11
inline bool isnan(float x) { return _isnan(x); }
inline bool isnan(double x) { return _isnan(x); }
inline bool isfinite(float x) { return _finite(x); }
inline bool isfinite(double x) { return _finite(x); }
}
}
#endif

File: src/google/protobuf/util/internal/utility.cc (hunk 1 above)
Comment (xfxyjwf) [inline src/google/protobuf/util/internal/utility.cc]: Could you change it to "MathLimits<double>::IsNan(value)"?

Comment (ostrovsky) [inline src/google/protobuf/util/internal/utility.cc]: It does not seem to work.  I can see in other places, such as strutil.cc
that IsNaN is locally defined.  Could that be a possibility:

inline bool IsNaN(double value) {
  // NaN is never equal to anything, even itself.
  return value != value;
}

Comment (xfxyjwf) [inline src/google/protobuf/util/internal/utility.cc]: From reading this thread:
http://stackoverflow.com/questions/570669/checking-if-a-double-or-float-is-nan-in-c
what I get is that "value != value" may not work reliable.

Could you elaborate why "MathLimits::IsNan(value)" doesn't work? To make the code portable we need a portable version of IsNan and I would prefer making the MathLimits one work for that purpose.

Comment (ostrovsky) [inline src/google/protobuf/util/internal/utility.cc]: Hello Feng Xiao,

you are right.  MathLimits IsNaN works fine.  I am made a silly mistake
earlier, which resulted in a compilation error.  It is corrected and
committed to GitHub.

Best regards,

Karol

**Comment by xfxyjwf** [comment]
The change looks good to me. Could you squash the commits into one commit?

**Comment by ostrovsky** [comment]
Hello Feng Xiao,

I went ahead and changed all the places with special IsNaN and isnan
implementations to MathLimits::IsNaN to make the code more portable.
Please, review the changes and I will squash the commits afterwards so that
you can merge them.

Best regards,

Karol

**Comment by ostrovsky** [comment]
Please, can someone restart the pull build that hanged?  There is no build
error there.

Thank you,

Karol

**Comment by xfxyjwf** [comment]
Looks good to me. Please squash the commits and I'll merge this PR. Thanks very much for fixing this!

**Comment by ostrovsky** [comment]
Squashed.  Thank you for a very good code review.

**Pull request merged**